### PR TITLE
feat(Kotlin): Add Kotlin code for the array and linked list chapter

### PR DESCRIPTION
**Pull request by curtishd**
Kotlin support for _Chapter Array and LinkedList_ and the rest will be added later.

### CODE REVIEW

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    /* 在数组中查找指定元素 */
+    fun find(nums: IntArray, target: Int): Int {
+        for (i in nums.indices) {
+            if (nums[i] == target) return i
+        }
+        return -1
+    }
+}
+
+fun main() {
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:72]: Add `/* Driver Code */` above this line

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
+/**
+ * File: array.kt
+ * Created Time: 2024-1-25
+ * Author: dcurtis
+ */
+package chapter_array_and_linkedlist
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:6]: Add an empty line between the file header and the code

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.concurrent.ThreadLocalRandom
+
+class array() {
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:10]: Please remove the class if it is not necessary.

In Java, every method should be enclosed in a class. But most programming languages do not require it.

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
+/**
+ * File: array.kt
+ * Created Time: 2024-1-25
+ * Author: dcurtis
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:4]: Recommending adding your email here

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ */
+package chapter_array_and_linkedlist
+
+import java.util.concurrent.ThreadLocalRandom
+
+class array() {
+    /* 随机访问元素 */
+    fun randomAccess(nums: IntArray): Int {
+        // 在区间 [0, nums.length) 中随机抽取一个数字
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:13]: nums.size

(Please replace the language-specific glossaries)

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        for (i in nums.indices) {
+            if (nums[i] == target) return i
+        }
+        return -1
+    }
+}
+
+fun main() {
+    /* 初始化数组 */
+    val array = IntArray(5)
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:74]: arr

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/**
+ * File: linked_list.kt
+ * Created Time: 2024-1-25
+ * Author:dcurtis
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:4]: Author: dcurtis

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 8]
```diff
@@ -0,0 +1,86 @@
+/**
+ * File: linked_list.kt
+ * Created Time: 2024-1-25
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:3]: 2024-01-25

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+ * Author:dcurtis
+ */
+package chapter_array_and_linkedlist
+
+import utils.ListNode
+import utils.printLinkedList
+
+/* 链表节点类 */
+class linked_list {
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:12]: Remove the class if it is not necessary

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        var h = head;
+        for (i in 0..<index) {
+            h = h?.next
+        }
+        return h;
+    }
+
+    /* 在链表中查找值为 target 的首个节点 */
+    fun find(head: ListNode?, target: Int): Int {
+        var index = 0;
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:38]: Is it correct to add `;` at the end of the line? Please unify the code formats

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    /* 删除链表的节点 n0 之后的首个节点 */
+    fun remove(n0: ListNode?) {
+        val p = n0?.next
+        val n1 = p?.next
+        n0?.next = n1
+    }
+
+    /* 访问链表中索引为 index 的节点 */
+    fun access(head: ListNode?, index: Int): ListNode? {
+        var h = head;
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:29]: Is it correct to add `;` at the end of the line? Please unify the code formats

**Comment by krahets** [comment]
Thanks for the PR! Please address the comments and fix all the similar issues(I didn't comment the same issue in `list.kt` and `my_list.kt`).

**Comment by krahets** [comment]
It seems like you delete the pull request template. Please see:

If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:

- [ ] I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
- [ ] I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
- [ ] The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

**Comment by curtishd** [comment]
I have modified the formatting, comments and others. Please take a look.

**Comment by krahets** [comment]
> It seems like you delete the pull request template. Please see:
> 
> If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:
> 
> * [ ]  I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
> * [ ]  I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
> * [ ]  The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

Please check this comment, thx!

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 12]
```diff
@@ -0,0 +1,101 @@
+/**
+ * File: array.kt
+ * Created Time: 2024-01-25
+ * Author: curtishd(1023632660@qq.com)
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:4]: curtishd (1023632660@qq.com)

(Add a space between the username and email)

Comment (curtishd) [reply codes/kotlin/chapter_array_and_linkedlist/array.kt:4]: alright

**Comment by curtishd** [comment]
Apologize for my carelessness, I would wonder how to attach the console output to this pull request?

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+/* 访问链表中索引为 index 的节点 */
+fun access(head: ListNode?, index: Int): ListNode? {
+    var h = head
+    for (i in 0..<index) {
+        h = h?.next
+    }
+    return h
+}
+
+/* 在链表中查找值为 target 的首个节点 */
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:35]: The code highlight style seems broken from this line. Is it a bug of the GitHub editor?

File: codes/kotlin/chapter_array_and_linkedlist/list.kt [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    nums.removeAt(3)
+    println("删除索引 3 处的元素，得到 nums = $nums")
+
+    /* 通过索引遍历列表 */
+    var count = 0
+    for (i in nums.indices) {
+        count += nums[i]
+    }
+
+    /* 直接遍历列表元素*/
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/list.kt:50]: /* 直接遍历列表元素 */

File: codes/kotlin/chapter_array_and_linkedlist/list.kt [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+ * File: list.kt
+ * Created Time: 2024-01-25
+ * Author: curtishd(1023632660@qq.com)
+ */
+
+package chapter_array_and_linkedlist
+
+/* Driver Code */
+fun main() {
+    // 注意数组的元素类型是 int[] 的包装类 Integer[]
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/list.kt:11]: `Integer[]` is the language-specific(Java) comment

File: codes/kotlin/chapter_array_and_linkedlist/my_list.kt [hunk 16]
```diff
@@ -0,0 +1,138 @@
+/**
+ * File: my_list.kt
+ * Created Time: 2024-01-25
+ * Author: curtishd(1023632660@qq.com)
+ */
+
+package chapter_array_and_linkedlist
+
+class MyList {
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/my_list.kt:9]: Miss the class heading comment

File: codes/kotlin/chapter_array_and_linkedlist/my_list.kt [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    private var arr: IntArray = intArrayOf() // 数组（存储列表元素）
+    private var capacity = 10 // 列表容量
+    private var size = 0 // 列表长度（当前元素数量）
+    private var extendRatio = 2 // 每次列表扩容的倍数
+
+    /* 主构造函数初始化 */
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/my_list.kt:15]: /* 构造方法 */

File: codes/kotlin/chapter_array_and_linkedlist/array.kt [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+/* 删除索引 index 处的元素 */
+fun remove(nums: IntArray, index: Int) {
+    // 把索引 index 之后的所有元素向前移动一位
+    for (i in index..<nums.size - 1) {
+        nums[i] = nums[i + 1]
+    }
+}
+
+/* 遍历数组*/
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/array.kt:50]: /* 遍历数组 */

File: codes/kotlin/chapter_array_and_linkedlist/list.kt [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    val nums1 = ArrayList<Int>(listOf(6, 8, 7, 10, 9))
+    nums.addAll(nums1)
+    println("将列表 nums1 拼接到 nums 之后，得到 nums = $nums")
+
+    /* 排序列表 */
+    nums.sort() //排序后,列表元素从小到大排列
+    println("排序列表后 nums = $nums")
+}
+
+
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/list.kt:65]: Remove the redundant empty line

File: codes/kotlin/chapter_array_and_linkedlist/linked_list.kt (hunk 13 above)
Comment (curtishd) [reply codes/kotlin/chapter_array_and_linkedlist/linked_list.kt:35]: I guess so.

File: codes/kotlin/chapter_array_and_linkedlist/list.kt (hunk 15 above)
Comment (curtishd) [reply codes/kotlin/chapter_array_and_linkedlist/list.kt:11]: ok

**Comment by krahets** [comment]
> Apologize for my carelessness, I would wonder how to attach the console output to this pull request?

No problem! You can either copy&paste the outputs or upload the screenshots. For example, #783

**Comment by curtishd** [comment]
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

 I've tested the code and ensured the outputs are the same as the outputs of reference codes.
 I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
 The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).
Test result:
屏幕截图 2024-02-02 193557
屏幕截图 2024-02-02 193618
屏幕截图 2024-02-02 193635
屏幕截图 2024-02-02 193658

**Review by krahets — APPROVED** [review]
Thanks for the PR. Happy new year!

**Pull request merged**